Your pending requests

GabrielVidal1 opened the pull request:
Fixes #247 

So I created a generic pending-request component from where I have 2 new component:
- group-pending-requests (most of the original code regarding 'groups you manage' pending requests)
- user-group-invitations (the new pending request part in the 'group you joined' page, that lists group invitations for the current user).

smadbe reviewed (changes requested):
As discussed, try to keep the child component limited to UI things and use EventEmitter to pass the action to the parent.

GabrielVidal1 commented on src/app/modules/shared-components/components/pending-request/pending-request.component.ts:
```diff
@@ -13,59 +13,50 @@ import {
 })
 export class PendingRequestComponent implements OnInit, OnChanges, OnDestroy {
-  @Input() groupId?: string;
   @Input() showSwitch = true;
```
the generic component doesn't need the groupId, it's only used for the pending requests when we click on a group in the 'group you manage' list, so I moved it in the corresponding component

GabrielVidal1 commented on src/app/modules/shared-components/shared-components.module.ts:
```diff
@@ -52,6 +52,7 @@ import { UserPipe } from 'src/app/shared/pipes/userDisplay';
 import { SubSectionComponent } from './components/sub-section/sub-section.component';
 import { AddContentComponent } from './components/add-content/add-content.component';
 import { FloatingSaveComponent } from './components/floating-save/floating-save.component';
+import { PendingRequestComponent } from './components/pending-request/pending-request.component';
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/group-pending-request/group-pending-request.component.ts:
```diff
@@ -0,0 +1,70 @@
+  templateUrl: './group-pending-request.component.html',
+  styleUrls: [ './group-pending-request.component.scss' ],
+})
+export class GroupPendingRequestComponent implements OnInit, OnChanges {
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/user-group-invitations/user-group-invitations.component.ts:
```diff
@@ -0,0 +1,34 @@
+export class UserGroupInvitationsComponent implements OnInit {
+
+  getRequestsFunction!: (sort: string[]) => Observable<GroupInvitation[]>;
+  processRequestsFunction!: (groupId: string, memberIds: string[], action: Action) => Observable<Map<string, any>>;
```
I removed it all together

GabrielVidal1 commented:
I did some major refactoring getting rid off the function passed to the generic pending-request component.

I moved some of the functions that were initially in this component to separate files for more readability.

One downside to doing that is that the logic, that is now outside the generic component isn't tested anymore (the already existing tests being for the generic pending-request component). For now I just removed these tests.

smadbe commented on src/app/modules/group/components/pending-request/pending-request.component.html:
```diff
@@ -98,11 +97,11 @@
               <span class="accept" [ngClass]="{disabled: selection.length === 0}" (click)="onAccept()">
                 <span>Accept</span>
-                <i class="fa fa-check" *ngIf="state !== 'accepting'; else loading"></i>
+                <i class="fa fa-check" *ngIf="state !== 'processing'; else loading"></i>
```
No need for this spinner while processing, but what prevent the "accept" to be clicked while processing?

GabrielVidal1 replied:
fixed

smadbe commented on src/app/modules/group/components/pending-request/pending-request.component.spec.ts:
```diff
@@ -1,276 +0,0 @@
-import { waitForAsync, ComponentFixture, TestBed } from '@angular/core/testing';
```
You can split the tests between the two components (pending joining requests and pending requests), it should not be that hard. These tests are quite useful as these component are a bit difficult to test.

GabrielVidal1 replied:
fixed

smadbe commented on src/app/modules/group/components/pending-request/pending-request.component.ts:
```diff
@@ -4,160 +4,40 @@ import {
+  @Input() state: 'fetching' | 'processing' | 'ready' | 'fetchingError' = 'fetching';
 
-  state: 'fetching' | 'accepting' | 'rejecting' | 'ready' |'fetchingError' = 'fetching';
+  @Output() refetch = new EventEmitter<string[]>();
```
If it is only used to sort, call it "sort", it is difficult to understand what is this list of string otherwise.

GabrielVidal1 replied:
fixed

smadbe commented on src/app/modules/group/components/pending-request/pending-request.component.ts:
```diff
@@ -172,33 +52,9 @@ export class PendingRequestComponent implements OnInit, OnChanges, OnDestroy {
-          this.processRequestError(err);
-        }
-      );
+    this.state = 'processing';
```
If is it is input, you are not supposed to change it in the component (and I supposed the changed is not passed to the parent component)

GabrielVidal1 replied:
fixed

smadbe reviewed (changes requested):
Already better 👍

GabrielVidal1 commented on src/app/modules/group/components/pending-request/request-processing.ts:
```diff
@@ -0,0 +1,36 @@
+import { PendingRequest } from 'src/app/modules/group/http-services/get-requests.service';
+import { Action } from 'src/app/modules/group/http-services/request-actions.service';
+
+export function parseResults(data: Map<string, any>[]): Result {
```
fixed

GabrielVidal1 commented on src/app/modules/group/helpers/response-toast.ts:
```diff
@@ -0,0 +1,42 @@
+  countSuccess: number;
+}
+
+export function displayResponseToast(messageService: MessageService, result: Result, verb: string, msg: string): void {
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/pending-join-request/pending-join-request.component.ts:
```diff
@@ -0,0 +1,125 @@
+        result => {
+          this.state = 'ready';
+
+          displayResponseToast(
+            this.messageService,
+            parseResults(result),
+            params.type === Action.Accept ? 'accept' : 'reject',
+            params.type === Action.Accept ? 'accepted' : 'declined'
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/pending-join-request/pending-join-request.component.ts:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  ngOnInit(): void {
+    if (!this.showSwitch) this.columns = [ groupColumn ].concat(this.columns);
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/pending-join-request/pending-join-request.component.ts:
```diff
@@ -0,0 +1,125 @@
+    { field: 'user.login', header: 'USER' },
+    { field: 'at', header: 'REQUESTED ON' },
+  ];
+  subgroupSwitchItems = [
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/pending-join-request/pending-join-request.component.ts:
```diff
@@ -0,0 +1,125 @@
+  styleUrls: [ './pending-join-request.component.scss' ],
+})
+export class PendingJoinRequestsComponent implements OnInit, OnChanges, OnDestroy {
+  @Input() groupId?: string;
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/pending-join-request/pending-join-request.component.ts:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  onProcessRequests(params: { data: PendingRequest[], type: Action }): void {
+    processRequests(
+      (groupId: string, memberIds: string[], action: Action) => this.requestActionService.processJoinRequest(groupId, memberIds, action),
+      params.type,
+      params.data
+    )
+      .subscribe(
```
fixed

GabrielVidal1 commented on src/app/modules/group/components/pending-join-request/pending-join-request.component.ts:
```diff
@@ -0,0 +1,125 @@
+  templateUrl: './pending-join-request.component.html',
+  styleUrls: [ './pending-join-request.component.scss' ],
+})
+export class PendingJoinRequestsComponent implements OnInit, OnChanges, OnDestroy {
```
fixed

smadbe commented on src/app/modules/group/components/user-group-invitations/user-group-invitations.component.ts:
```diff
@@ -0,0 +1,89 @@
+    { field: 'at', header: 'REQUESTED ON' },
+  ];
+
+  state: 'fetching' | 'processing' | 'ready' | 'fetchingError' = 'fetching';
```
You never put the state to 'fetching' ?🤔
And as I initially suggested, if you do not use the different between the fetching and the processing state, use only one for both.

GabrielVidal1 replied:
The 'fecthing' state is set in the subscribe of the `dataFetching` Subject via the object returned by the `fetchingState()` function

``` typescript
this.dataFetching.pipe(
      switchMap(params =>
        merge(
          of(fetchingState()), // returns { tag: 'fetching' }
          this.getRequestsService.getGroupInvitations(params.sort)
            .pipe(map(readyState))
        )
      )
    ).subscribe(
      state => {
        this.state = state.tag; // set this.state to 'fetching'
        ...
```

smadbe replied:
Ok, sorry, I searched for the 'fetching' word. 
And for the second remark?  I see it used on line 12 of pending-join-request html but actually I don't think this disable state is still useful now we do the switchmap.

GabrielVidal1 replied:
fixed

smadbe commented on src/app/modules/group/http-services/request-actions.service.ts:
```diff
@@ -4,6 +4,7 @@ import { ActionResponse, successData, objectToMap } from 'src/app/shared/http-se
 import { map } from 'rxjs/operators';
 import { Observable } from 'rxjs';
 import { appConfig } from 'src/app/shared/helpers/config';
+import { Result } from '../helpers/response-toast';
```
A service cannot depend on response toast things :-/ But you can return `{ countRequests: int, countSuccess: int }` without importing it.

GabrielVidal1 replied:
fixed

smadbe commented on src/app/modules/group/http-services/request-actions.service.ts:
```diff
@@ -4,7 +4,7 @@ import { ActionResponse, successData, objectToMap } from 'src/app/shared/http-se
 import { Observable } from 'rxjs';
 import { appConfig } from 'src/app/shared/helpers/config';
-import { Result } from '../helpers/response-toast';
+import { Result } from '../components/pending-request/pending-request-response-handling';
```
A service should not depend on a component.

GabrielVidal1 replied:
fixed

smadbe reviewed (approved):
I'll not re-review everything, I'll assume that's ok.
